packages/core/src/loaders/progressiveLoader.ts

```diff
@@ -0,0 +1,382 @@
+      }
+    }
+  }
+
+  return nearby;
+}
+
+/** Actually fills the nearby frames from the given frame */
+function fillNearbyFrames(
+  listener: ProgressiveListener,
+  imageStatus: Map<string, ImageStatus>,
+  request,
+  image,
+  options
+) {
+  if (!request?.nearbyRequests?.length) {
+    return;
+  }
+
+  const {
+    arrayBuffer,
+    offset: srcOffset,
+    type,
+    length: frameLength,
+  } = options.targetBuffer;
+  if (!arrayBuffer || srcOffset === undefined || !type) {
+    return;
+  }
+  const scalarData = new Float32Array(arrayBuffer);
+  const bytesPerPixel = scalarData.byteLength / scalarData.length;
+  const offset = options.targetBuffer.offset / bytesPerPixel; // in bytes
+  // since set is based on the underlying type,
+  // we need to divide the offset bytes by the byte type
+  const src = scalarData.slice(offset, offset + frameLength);
+
+  for (const nearbyItem of request.nearbyRequests) {
+    try {
+      const { itemId: targetId, status } = nearbyItem;
+      const targetStatus = imageStatus.get(targetId);
+      if (targetStatus !== undefined && targetStatus < status) {
+        continue;
+      }
+      const targetOptions = listener.getTargetOptions(targetId);
+      const { offset: targetOffset } = targetOptions.targetBuffer as any;
+      scalarData.set(src, targetOffset / bytesPerPixel);
+      const nearbyImage = {
+        ...image,
+        status,
+      };
+      listener.successCallback(targetId, nearbyImage, status);
+      imageStatus[targetId] = status;
+    } catch (e) {
+      console.log("Couldn't fill nearby item ", nearbyItem.itemId, e);
+    }
+  }
+}
+
+function addStageStatus(stageStatus: Map<string, StageStatus>, stage) {
+  const { id } = stage;
+  const status = stageStatus.get(id) || {
+    stageId: id,
+    startTime: Date.now(),
+    stageStartTime: null,
+    numberOfImages: 0,
+    numberOfFailures: 0,
+    remaining: 0,
+  };
+  status.remaining++;
+  stageStatus.set(id, status);
+  return status;
+}
+
+function updateStageStatus(
+  stageStatus: Map<string, StageStatus>,
+  stage,
+  failure?,
+  skipped = false
+) {
+  const { id } = stage;
+  const status = stageStatus.get(id);
+  if (!status) {
+    console.warn('Stage already completed:', id);
+    return;
+  }
+  status.remaining--;
+  if (failure) {
+    status.numberOfFailures++;
+  } else if (!skipped) {
+    status.numberOfImages++;
+  }
+  if (!skipped && !status.stageStartTime) {
+    status.stageStartTime = Date.now();
+  }
+  if (!status.remaining) {
+    const {
+      numberOfFailures,
+      numberOfImages,
+      stageStartTime = Date.now(),
+      startTime,
+    } = status;
+    const detail: EventTypes.ImageLoadStageEventDetail = {
+      stageId: id,
+      numberOfFailures,
+      numberOfImages,
+      stageDurationInMS: stageStartTime ? Date.now() - stageStartTime : null,
+      startDurationInMS: Date.now() - startTime,
+    };
+    triggerEvent(eventTarget, Events.IMAGE_LOAD_STAGE, detail);
```

Comment:
I think IMAGE_RETRIEVAL_STAGE makes more sense given all the other mentions of the retrieve in this context, wdyt